src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -200,20 +200,15 @@ async Task EnsureLong()
                 // We have a special case when reading async notifications - a timeout may be normal
                 // shouldn't be fatal
                 // Note that mono throws SocketException with the wrong error (see #1330)
-                catch (IOException e) when (
-                    dontBreakOnTimeouts && (e.InnerException as SocketException)?.SocketErrorCode ==
-                       (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock)
-                )
+                catch (Exception e) when (e is OperationCanceledException ||
```

Comment:
Pure styling: I'd simply introduce a switch under the catch-all below - there's very little value in different catch blocks here.

I'd also introduce Debug.Assert for async when catching OperationCanceledException, and for !async when catching IOException with timeout - mostly as a means of making more explicit what the code does.